lib/Common/Codex/Utf8Helper.h

```diff
@@ -86,31 +86,24 @@ namespace utf8
         return WideStringToNarrow(Allocator::allocate, sourceString, sourceCount, destStringPtr, destCount, allocateCount);
     }
 
-    ///
-    /// Use the codex library to encode a UTF8 string to UTF16.
-    /// The caller is responsible for freeing the memory, which is allocated
-    /// using Allocator.
-    /// The returned string is null terminated.
-    ///
-    template <typename AllocatorFunction>
-    HRESULT NarrowStringToWide(_In_ AllocatorFunction allocator,_In_ LPCSTR sourceString, size_t sourceCount, _Out_ LPWSTR* destStringPtr, _Out_ size_t* destCount, size_t* allocateCount = nullptr)
+    inline HRESULT NarrowStringToWideNoAlloc(_In_ LPCSTR sourceString, size_t sourceCount,
+        __out_ecount(destBufferCount) LPWSTR destString, size_t destBufferCount, _Out_ size_t* destCount)
     {
-        size_t cbSourceString = sourceCount;
         size_t sourceStart = 0;
-        size_t cbDestString = (sourceCount + 1) * sizeof(WCHAR);
-        if (cbDestString < sourceCount) // overflow ?
+        size_t cbSourceString = sourceCount;
+
+        if (sourceCount >= MAXUINT32)
         {
+            destString[0] = WCHAR(0);
             return E_OUTOFMEMORY;
         }
 
-        WCHAR* destString = (WCHAR*)allocator(cbDestString);
-        if (destString == nullptr)
+        if (destString == nullptr || sourceCount >= destBufferCount)
```

Comment:
If `destString == nullptr` then we shouldn't be writing to it at all; perhaps put that check first?

Reply:
Also isn't `sourceCount >= destBufferCount` legal? E.g. if the input narrow string has more 3-byte utf8 codepoints than 1-bye utf8 codeopints, the utf16 representation can be shorter than the input.

Reply:
I didn't understand your first comment.

Second comment; -> sourceCount represents the length of `char16` buffer. Allocator is supposed to allocate equal or more (prior to calculating exact correspondence). So we can convert using the fast path below.

Reply:
This condition is saying `if (destString == nullptr || ...) { destString[0] = 0}` which will (try to) write to the null pointer. Also the previous condition will write to `destString` without checking whether it is null. We should either be checking if `destString == nullptr` up front and exiting with an error without writing to it, or we should be relying on SAL to make sure that we never pass in a null pointer I think.

Regarding the second point; Requiring that `sourceCount >= destBufferCount` is safe, but it may lead to excess memory usage. Prior to this PR the over-estimating allocations are (I believe) always temporary, and the string is copied to a more accurately sized buffer for long-term storage, or freed once the variable goes out of scope. The new usage that you're adding in recycler-allocates a (potentially) over-estimated amount and hands that off to to be stored in a `LiteralString`.

Perhaps it is worth being clear about what the contract of this function is? Are you supposed to supply a buffer with `destBufferCount == sourceCount`, and then the caller is responsible for adjusting to `destCount` when the call returns? Is the caller supposed to supply a buffer with size at least `destCount` which is initially unknown, and it is an error if `destBufferCount < destCount`?

Reply:
Second part; Yes, the idea is/was, instead of double pass, allocate ASCII amount of memory. OR we will endup having double pass for all the strings.